File: chapters/03_session_and_dm_scope/v2_main_session.ts
// TypeScript mirror for Chapter 03: Session and DM Scope (v2)
// Run with: npx ts-node v2_main_session.ts

class SessionStore {
  private _messages: Map<string, string[]> = new Map();

  append(key: string, message: string): void {
    if (!this._messages.has(key)) {
      this._messages.set(key, []);
    }
    this._messages.get(key)!.push(message);
  }

  history(key: string): string[] {
    return [...(this._messages.get(key) || [])];
  }
}

function mainSessionKey(workspace: string, agent: string): string {
  return `${workspace}:${agent}:main`;
}

function handleWithMainSession(
  store: SessionStore,
  text: string,
  workspace: string = "demo",
  agent: string = "openclaw"
): { session_key: string; history_size: number } {
  const key = mainSessionKey(workspace, agent);
  store.append(key, `user:${text}`);
  store.append(key, `assistant:ack:${text}`);
  return {
    session_key: key,
    history_size: store.history(key).length,
  };
}

function main(): void {
  const store = new SessionStore();
  console.log(handleWithMainSession(store, "hello"));
}

main();

File: chapters/03_session_and_dm_scope/v2_main_session.py
from __future__ import annotations


class SessionStore:
    def __init__(self) -> None:
        self._messages: dict[str, list[str]] = {}

    def append(self, key: str, message: str) -> None:
        self._messages.setdefault(key, []).append(message)

    def history(self, key: str) -> list[str]:
        return list(self._messages.get(key, []))


def main_session_key(workspace: str, agent: str) -> str:
    return f"{workspace}:{agent}:main"


def handle_with_main_session(store: SessionStore, text: str, workspace: str = "demo", agent: str = "openclaw") -> dict[str, object]:
    key = main_session_key(workspace, agent)
    store.append(key, f"user:{text}")
    store.append(key, f"assistant:ack:{text}")
    return {
        "session_key": key,
        "history_size": len(store.history(key)),
    }


def main() -> None:
    store = SessionStore()
    print(handle_with_main_session(store, "hello"))


if __name__ == "__main__":
    main()
